src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.ReadHelper.cs

```diff
@@ -101,29 +114,58 @@ public partial class JsonTypeInfo<T>
 
         internal sealed override async ValueTask<object?> DeserializeAsObjectAsync(Stream utf8Json, CancellationToken cancellationToken)
         {
-            T? result = await DeserializeAsync(utf8Json, cancellationToken).ConfigureAwait(false);
+            // Note: The ReadBufferState ctor rents pooled buffers.
+            StreamReadBufferState bufferState = new StreamReadBufferState(utf8Json, Options.DefaultBufferSize);
+            T? result = await DeserializeAsync(bufferState, cancellationToken).ConfigureAwait(false);
+            return result;
+        }
+
+        internal sealed override async ValueTask<object?> DeserializeAsObjectAsync(PipeReader utf8Json, CancellationToken cancellationToken)
+        {
+            T? result = await DeserializeAsync(new PipeReadBufferState(utf8Json), cancellationToken).ConfigureAwait(false);
             return result;
         }
 
         internal sealed override object? DeserializeAsObject(Stream utf8Json)
             => Deserialize(utf8Json);
 
-        internal bool ContinueDeserialize(
-            ref ReadBufferState bufferState,
+        internal bool ContinueDeserialize<TReadBufferState>(
+            ref TReadBufferState bufferState,
             ref JsonReaderState jsonReaderState,
             ref ReadStack readStack,
             out T? value)
+            where TReadBufferState : struct, IReadBufferState<TReadBufferState>
         {
-            var reader = new Utf8JsonReader(bufferState.Bytes, bufferState.IsFinalBlock, jsonReaderState);
-            bool success = EffectiveConverter.ReadCore(ref reader, out value, Options, ref readStack);
+            Utf8JsonReader reader;
+            if (bufferState.Bytes.IsSingleSegment)
+            {
+                reader = new Utf8JsonReader(
+#if NET
+                    bufferState.Bytes.FirstSpan,
```

Comment:
What necessitates specializing to the `FirstSpan` property in core?

Reply:
I had a vague memory that it was faster.

https://github.com/dotnet/runtime/issues/27730